## Import required library

In [2]:
import pandas as pd

In [3]:
Population=pd.read_csv('PEA11.20231021T221005.csv')

## DATA EXPLORATION

Find out the shape of the read CSV to get an idea of size

In [4]:
Population.shape
print('Number of Rows and Columns respectively:',Population.shape)

Number of Rows and Columns respectively: (11817, 6)


In [5]:
Population.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11817 entries, 0 to 11816
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   STATISTIC Label     11817 non-null  object
 1   Year                11817 non-null  int64 
 2   Single Year of Age  11817 non-null  object
 3   Sex                 11817 non-null  object
 4   UNIT                11817 non-null  object
 5   VALUE               11817 non-null  int64 
dtypes: int64(2), object(4)
memory usage: 554.1+ KB


The dataset, above, was checked to detect any missing/blank values. The number of non-null values match the number of rows
so no missing values detected. 'info() was also used to check the data type of each column in the dataset. This can be later seen when a section of the dataset is displayed

To confirm this, the sum of the blank values are counted and collated for each column as below

In [6]:
Population.isna().sum()

STATISTIC Label       0
Year                  0
Single Year of Age    0
Sex                   0
UNIT                  0
VALUE                 0
dtype: int64

The above only checks for missing value, but sometimes due to data entry, the inputted values could be invalid even if filled in. The code below will utilise a function to check for a list of commonly entered values that wouldn't count as blank but also wouldn't be valid entries so it would be best to replace with 'NaN'

For the function 'count_invalid_entries'



In [22]:
# List of values to search for
invalid_entries = ['n.a', '-','--', 'NA', 'Not Applicable', 'n/a']

#This will now count all the rows that match the list of commonly entered values in place of blanks in data for all columns
#'.lower()'' is used to change both assessed entries and given invalid entries to lowercase to make the function 
#non -sensitive
def count_invalid_entries(dataframe, invalid_entries):
    matching_rows = dataframe[dataframe.apply(lambda row: any(item.lower() in row.values for item in invalid_entries), axis=1)]
    return len(matching_rows)

# Call the function and count matching rows in all columns
count = count_invalid_entries(Population, invalid_entries)
print(f"Number of rows that match the list of strings in all columns: {count}")

Number of rows that match the list of strings in all columns: 0


The function returns no values matching the list inside the whole dataset

Now that missing and invalid values have been confirmed above, duplicate values are detected below and the same idea as above, is used to display the sum of any rows that are similar for **ALL** values in the row

In [17]:
Population.duplicated().sum()
print ('The sum of duplicate rows in the dataset is:',Population.duplicated().sum(),'rows')

The sum of duplicate rows in the dataset is: 0 rows


Below, a snapshot into the first ('.head(5)') and last ('.tail(5)) 5 rows is displayed

This is useful as all column names are displayed at once

In [9]:
Population.head(5)

STATISTIC Label  Year Single Year of Age         Sex  \
0  Population estimates from 1926  1926           All ages  Both sexes   
1  Population estimates from 1926  1926           All ages        Male   
2  Population estimates from 1926  1926           All ages      Female   
3  Population estimates from 1926  1926       Under 1 year  Both sexes   
4  Population estimates from 1926  1926       Under 1 year        Male   

     UNIT    VALUE  
0  Number  2971992  
1  Number  1506889  
2  Number  1465103  
3  Number    55391  
4  Number    28084

In [8]:
Population.tail(5)

STATISTIC Label  Year Single Year of Age         Sex  \
11812  Population estimates from 1926  2023           98 years        Male   
11813  Population estimates from 1926  2023           98 years      Female   
11814  Population estimates from 1926  2023  99 years and over  Both sexes   
11815  Population estimates from 1926  2023  99 years and over        Male   
11816  Population estimates from 1926  2023  99 years and over      Female   

         UNIT  VALUE  
11812  Number    197  
11813  Number    632  
11814  Number   1600  
11815  Number    352  
11816  Number   1248

From above, it is good to note that 11816 is the designation of the last row in the dataset. This is important as the rows are read and displayed as integer index unless otherwise assigned

It is also seen that 'VALUE' and 'Year' are the only numerical columns above

This is also corroborated below when '.describe()' was used but '.describe(include=object)' had to be used to show the
remaining categorical columns

In [18]:
Population.describe()

Year         VALUE
count  11817.000000  1.181700e+04
mean    1996.410256  5.352461e+04
std       24.463165  2.867754e+05
min     1926.000000  6.000000e+00
25%     1986.000000  1.242700e+04
50%     2004.000000  2.667500e+04
75%     2014.000000  3.557700e+04
max     2023.000000  5.281612e+06

As the columns above are numerical values, it is possible to compute some statistics based on the datset

In [12]:
Population.describe(include=object)

STATISTIC Label Single Year of Age         Sex    UNIT
count                            11817              11817       11817   11817
unique                               1                101           3       1
top     Population estimates from 1926           All ages  Both sexes  Number
freq                             11817                117        3939   11817

## DATA PREPARATION

From above, it is seen that the columns 'UNIT' and 'STATISTIC LABEL' have only 1 unique value

To make the data more streamlined, the columns with only 1 unique value are dropped as the lack of variation in data would not give much useful info during the analysis.

In [25]:
Population_df=Population.drop(columns=['UNIT','STATISTIC Label'],axis=1,inplace=False)

'inplace' is set to False instead of True so a new variable can be assigned to the modified datafram instead of mutating the data in place <br><br>
Setting it to False was selected majorly because it would be useful to track any intermediate dataframes, when debugging, created as part of the data preparation process <br><br>
The new shape and first 5 rows of the new modified dataframe is displayed to see the results

In [26]:
Population_df
print('Number of Rows and Columns respectively:',Population_df.shape)
Population_df.head()

Number of Rows and Columns respectively: (11817, 4)


Year Single Year of Age         Sex    VALUE
0  1926           All ages  Both sexes  2971992
1  1926           All ages        Male  1506889
2  1926           All ages      Female  1465103
3  1926       Under 1 year  Both sexes    55391
4  1926       Under 1 year        Male    28084

The columns were modified to make more visually readable
The columns 'Single Year of Age' was chnaged to 'Age' and 'VALUE' changed to 'Value' for simplicity and to align with other columns

Note: 'inplace' was set to True as the change in reuslting data is minimal

In [31]:
Population_df.rename(columns={'Single Year of Age': 'Age', 'VALUE': 'Value'}, inplace=True)

The first 5 rows of the newly modified dataset are shown to display the changes

In [35]:
Population_df.head()

Year           Age         Sex    Value
0  1926      All ages  Both sexes  2971992
1  1926      All ages        Male  1506889
2  1926      All ages      Female  1465103
3  1926  Under 1 year  Both sexes    55391
4  1926  Under 1 year        Male    28084

This section will now look to create a number of sub dataframes to give the data focus on certain aspects. Each year, for each age, is reported with numbers for 'Male', 'Female' and 'Both sexes'
Using a function, 4 different datsets will be created separated by sex:
* Both Sexes: Just the total of femal and male per year
* Male: Only Male population per year
* Female: Only Female population per year
* Male_Female: Population per year without totalling it up using Both sexes

In [39]:
#initialising the function
def split_by_sex(dataframe):
    Both_sexes_df = dataframe[dataframe['Sex'] == 'Both sexes']
    Male_df = dataframe[dataframe['Sex'] == 'Male']
    Female_df = dataframe[dataframe['Sex'] == 'Female']
    Male_Female_df = dataframe[dataframe['Sex'].isin(['Male', 'Female'])]
    return Both_sexes_df, Male_df, Female_df, Male_Female_df

Both_sexes_df, Male_df, Female_df, Male_Female_df = split_by_sex(Population_df)

First 5 rows of all new datasets are displayed below

In [40]:
Both_sexes_df.head()

Year           Age         Sex    Value
0   1926      All ages  Both sexes  2971992
3   1926  Under 1 year  Both sexes    55391
6   1926        1 year  Both sexes    55876
9   1926       2 years  Both sexes    58273
12  1926       3 years  Both sexes    59050

In [41]:
Male_df.head()

Year           Age   Sex    Value
1   1926      All ages  Male  1506889
4   1926  Under 1 year  Male    28084
7   1926        1 year  Male    28374
10  1926       2 years  Male    29728
13  1926       3 years  Male    29872

In [42]:
Female_df.head()

Year           Age     Sex    Value
2   1926      All ages  Female  1465103
5   1926  Under 1 year  Female    27307
8   1926        1 year  Female    27502
11  1926       2 years  Female    28545
14  1926       3 years  Female    29178

In [44]:
Male_Female_df.head()

Year           Age     Sex    Value
1  1926      All ages    Male  1506889
2  1926      All ages  Female  1465103
4  1926  Under 1 year    Male    28084
5  1926  Under 1 year  Female    27307
7  1926        1 year    Male    28374

## DATA VISUALISATION

To start the visualisations, the matplot library will be imported


In [52]:
import matplotlib.pyplot as plt

In [ ]:
The trend of 